packages/@react-aria/utils/src/mergeProps.ts

```diff
@@ -18,42 +18,66 @@ interface Props {
   [key: string]: any
 }
 
+// taken from: https://stackoverflow.com/questions/51603250/typescript-3-parameter-list-intersection-type/51604379#51604379
+type TupleTypes<T> = { [P in keyof T]: T[P] } extends { [key: number]: infer V }
+  ? V
+  : never;
+type UnionToIntersection<U> = (U extends any
+  ? (k: U) => void
+  : never) extends ((k: infer I) => void)
+  ? I
+  : never;
+
 /**
  * Merges multiple props objects together. Event handlers are chained,
  * classNames are combined, and ids are deduplicated. For all other props,
- * the second props object overrides the first.
+ * the last prop object overrides all previous ones.
  * @param a - The first set of props to merge.
- * @param b - The second set of props to merge.
+ * @param rest - The remaining sets of props to merge.
  */
-export function mergeProps<T extends Props, U extends Props>(a: T, b: U): T & U {
-  let res: Props = {};
-  for (let key in a) {
-    // Chain events
-    if (/^on[A-Z]/.test(key) && typeof a[key] === 'function' && typeof b[key] === 'function') {
-      res[key] = chain(a[key], b[key]);
-
-    // Merge classnames, sometimes classNames are empty string which eval to false, so we just need to do a type check
-    } else if (key === 'className' && typeof a.className === 'string' && typeof b.className === 'string') {
-      res[key] = classNames(a.className, b.className);
-
-    } else if (key === 'UNSAFE_className' && typeof a.UNSAFE_className === 'string' && typeof b.UNSAFE_className === 'string') {
-      res[key] = classNames(a.UNSAFE_className, b.UNSAFE_className);
+export function mergeProps<T extends Props, U extends Props[]>(
+  a: T,
+  ...rest: U
```

Comment:
Does this mean all of the subsequent props objects after the first one must be the same type, or does it allow differences as long as they are all compatible with `Props`? If so, do we still need to treat `a` separately or would it make sense to make them all part of the same rest parameter?

Reply:
They can all be different. I probably should have added a test that showed that, but here's sandbox I have up with an example: https://codesandbox.io/s/mergeprops-irpn6?file=/src/index.ts

At first I did just spread all of the args, but then I separated `a` out just to start off the `result` variable a little more easily, but I suppose that's not actually super necessary. Maybe it would be better to switch it to just being `...rest` or `...props` or something.

Reply:
Alright, I went ahead and got rid of a and just have `...args` passed in. I also added another test to show that it will allow you to add other props in the later objects and it will merge them altogether.